lib/FIR/afforestation.cc

```diff
@@ -1057,27 +309,60 @@ class FortranIRLowering {
 
   Program *program() { return fir_; }
 
+  // convert a parse tree data reference to an Expression
+  template<typename A> Expression ToExpression(const A &a) {
+    return {std::move(semantics::AnalyzeExpr(semanticsContext_, a).value())};
+  }
+
+  // build a simple arithmetic Expression
+  template<template<typename> class OPR>
+  Expression ConsExpr(Expression e1, Expression e2) {
+    evaluate::ExpressionAnalyzer context{semanticsContext_};
+    ConformabilityCheck(context.GetContextualMessages(), e1, e2);
+    return evaluate::NumericOperation<OPR>(context.GetContextualMessages(),
+        std::move(e1), std::move(e2),
+        context.GetDefaultKind(common::TypeCategory::Real))
+        .value();
+  }
+  Expression ConsExpr(
+      common::RelationalOperator op, Expression e1, Expression e2) {
+    evaluate::ExpressionAnalyzer context{semanticsContext_};
+    return evaluate::AsGenericExpr(evaluate::Relate(
+        context.GetContextualMessages(), op, std::move(e1), std::move(e2))
+                                       .value());
+  }
+  parser::Name MakeTemp(Type tempType) {
+    auto name{NewTemporaryName()};
+    auto details{semantics::ObjectEntityDetails{true}};
+    details.set_type(std::move(*tempType));
+    auto *sym{&semanticsContext_.globalScope().MakeSymbol(
+        name, {}, std::move(details))};
```

Comment:
If we ever try to refer to this symbol in a message, it won't work because the name is not part of the original cooked character stream so there is no provenance information. Is that a problem?

Reply:
Our working premise is that semantics checking will have been fully completed, so hopefully issuing user input sorts of diagnostic messages is also (mostly?) completed.  Still, who can guarantee there will be no further messages?  As to not having provenance information, that seems like a given here. The "symbol" is made up by the compiler so it doesn't make sense for it to have any sort of context in the input source.

Reply:
Semantics will have caught any fatal errors, but optimization can generate informational messages.  Any `CharBlock` in the symbol table should be mappable by the provenance system.   Maybe the symbol table should `CHECK()` that if it doesn't already. There is already a means for storing compiler-generated text there (it comes up when the prescanner or preprocessor has to insert a newline or spaces that weren't originally present).

Reply:
@klausler I didn't know about that feature in the prescanner. Can you provide a pointer to this? Thanks.

Reply:
Not for this check-in but something to think about:  We're in agreement that any errors have been reported before the CFG pass starts; however, like Peter says, the later analyses might offer information to the user so we have to anticipate that messages will pop out that are related to compiler-generated names.  We should think about a way to make these meaningful to the user (instead of, for example, trying to pattern match on such names & keeping the messages quiet).

Reply:
+1